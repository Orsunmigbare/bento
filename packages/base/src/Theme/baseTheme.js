import { ThemeManager } from "@staccx/theme"

export const theme = {
  name: "Default theme",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px",
    micro: "3px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px",
    breakout: "640px"
  },
  color: {
    white: "#fff",
    bg: "#fff",
    subtleHover: "#F7F7FF",
    bgGray: "#f7f7ff",
    line: "#EBEBF2",
    gray: "#767676",
    wcag: "#777",
    disabled: "#BABABA",
    primary: "#00a190",
    secondary: "#F5A623",
    text: "#2b2b2b",
    negative: "#ea526f",
    positive: "#4A90E2",
    warning: "#c54059",
    black: "#000",
    green: "#00a190" // for toggle eks
  },
  fontFamily: {
    heading: "sans-serif",
    body: "sans-serif"
  },
  font: {
    h1: "36px",
    h2: "26px",
    h3: "22px",
    h4: "18px",
    h5: "16px",
    h6: "14px",
    base: "14px",
    tiny: "12px",
    input: "16px",
    huge: "44px"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px"
}

const manager = new ThemeManager(theme)
export default manager
