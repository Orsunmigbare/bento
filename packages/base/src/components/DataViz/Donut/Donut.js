import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import {
  color,
  themify
} from "@staccx/theme"

const Donut = ({ percentage, className, ...rest }) => {
  const dash = 57
  const dashoffset = 57 * percentage - 57
  return (
    <Wrapper
      viewBox="0 0 24 24"
      width="24"
      height="24"
      className={className}
      {...rest}
    >
      <title>{percentage * 100}%</title>
      <g fill="none" fillRule="evenodd" strokeWidth="6">
        <Bg cx="12" cy="12" r="9" />
        <Value
          cx="12"
          cy="12"
          r="9"
          transform="rotate(-90 12 12)"
          dash={dash}
          dashoffset={dashoffset}
        />
      </g>
    </Wrapper>
  )
}

export const DONUT = "donut"

const Wrapper = styled.svg`
  transform: rotateY(180deg);
  ${themify(DONUT)};
`

export const DONUT_BG = "donut_bg"

const Bg = styled.circle`
  stroke: ${color.line};
  ${themify(DONUT_BG)};
`

export const DONUT_VALUE = "donut_value"

const Value = styled.circle`
  stroke: ${color.primary};
  stroke-dasharray: ${p => p.dash};
  stroke-dashoffset: ${p => p.dashoffset};
  ${themify(DONUT_VALUE)};
`

Donut.defaultProps = {
  className: ""
}

Donut.propTypes = {
  percentage: PropTypes.number.isRequired,
  className: PropTypes.string
}

export default Donut
