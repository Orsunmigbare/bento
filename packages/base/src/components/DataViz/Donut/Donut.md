```jsx
<Donut percentage={0.25} />
```
