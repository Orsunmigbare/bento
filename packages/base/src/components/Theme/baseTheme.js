export const theme = {
  name: "StaccXBase",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px"
  },
  color: {
    bg: "#fff",
    bgGray: "#f7f7ff",
    primary: "#00a190",
    gray: "#767676",
    text: "#2b2b2b",
    white: "#fff",
    red: "#ea526f",
    green: "#00a190",
    blue: "#988EFB",
    yellow: "#f4da87",
    line: "#EBEBF2",
    wcag: "#777",
    disabled: "#BABABA",
    warning: "#c54059",
    black: "#000",
    subtleHover: "#F7F7FF"
  },
  font: {
    size: {
      h1: "36px",
      h2: "26px",
      h3: "22px",
      h4: "18px",
      h5: "16px",
      h6: "14px",
      base: "14px",
      tiny: "12px",
      input: "16px",
      huge: "44px"
    },
    heading: "Libre Baskerville",
    body: "Libre Franklin"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px"
}

export default theme
