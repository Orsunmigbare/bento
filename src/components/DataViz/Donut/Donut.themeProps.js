import { themePropTypes } from "../../../theming"

const themeProps = {
  donut: {
    name: "donut",
    description: "Wrapper style for donut",
    type: themePropTypes.style
  },
  donutBackground: {
    name: "donut_bg",
    description: "Circle style for donut background",
    type: themePropTypes.style
  },
  donutValue: {
    name: "donut_value",
    description: "Value style for donut",
    type: themePropTypes.style
  }
}
export default themeProps
