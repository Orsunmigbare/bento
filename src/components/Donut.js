import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"

const Donut = ({ percentage }) => {
  const dash = 57
  const dashoffset = 57 * percentage
  console.log(dashoffset)
  return (
    <Graph viewBox="0 0 24 24" width="24" height="24">
      <g fill="none" fillRule="evenodd" strokeWidth="6">
        <Bg cx="12" cy="12" r="9" />
        <Value
          cx="12"
          cy="12"
          r="9"
          transform="rotate(-90 12 12)"
          dash={dash}
          dashoffset={dashoffset}
        />
      </g>
    </Graph>
  )
}

const Graph = styled.svg`
  transform: rotateY(180deg);
`

const Bg = styled.circle`
  stroke: ${p => p.theme.color.primary};
`

const Value = styled.circle`
  stroke: ${p => p.theme.color.line};
  stroke-dasharray: ${p => p.dash};
  stroke-dashoffset: ${p => p.dashoffset};
`

Donut.propTypes = {
  percentage: PropTypes.number.isRequired
}

export default Donut
