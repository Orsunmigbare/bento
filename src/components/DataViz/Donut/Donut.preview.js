import React from "react"
import Donut from "./Donut"
import Slider from "../../Forms/Slider/Slider/Slider"
import Label from "../../Forms/Label/Label"
import Wrapper from "../../Layout/Wrapper/Wrapper"

const preview = {
  title: "Donut",
  category: "Components/Dataviz",
  tags: ["component", "dataviz"],
  component: Donut,
  render: props => <Donut progress={0.25} {...props} />,
  renderExampleController: ({ setComponentState }) => (
    <Wrapper>
      <Label>Progress</Label>
      <Slider
        name={"DonutExample"}
        min={0}
        max={1}
        value={0.25}
        step={0.01}
        onChange={e =>
          setComponentState({ progress: parseFloat(e.target.value) })
        }
      />
      <Label>Warning threshold</Label>
      <Slider
        name={"warningThreshold"}
        min={0}
        max={1}
        value={0.9}
        step={0.01}
        onChange={e =>
          setComponentState({
            warningThreshold: parseFloat(e.target.value)
          })
        }
      />
    </Wrapper>
  )
}

export default preview
