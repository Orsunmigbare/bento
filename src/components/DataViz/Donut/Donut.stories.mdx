import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Donut from "./Donut.js"

<Meta title="Components|DataViz/Donut" component={Donut}/>

# Donut

Donut displays progress from 0 to 100%, as a radial progressbar

```javascript
import {Donut} from "@staccx/bento"
```

## Donut with progress 0%

<Preview>
  <Story name="0%">
   <Donut progress="0"/>
  </Story>
</Preview>

   
## Donut with progress 50%
<Preview>
  <Story name="50%">
   <Donut progress="0.5"/>
  </Story>
</Preview>

## Donut with progress 75%
<Preview>
  <Story name="75%">
   <Donut progress="0.75"/>
  </Story>
</Preview>



## Donut with progress 100%
<Preview>
  <Story name="100%">
   <Donut progress="1"/>
  </Story>
</Preview>

## Donut with threshold set to 10%
<Preview>
    <Story name="Threshold low">
       <Donut progress="0.2" warningThreshold="0.1"/>
     </Story>
</Preview>




## Different possible pitfalls 
These are examples showing what happends if the component recieves props that are
out of range

## Donut with progress over 100%

<Preview>
  <Story name="110%">
   <Donut progress="1.1"/>
  </Story>
</Preview>


## Donut with progress under 0%

<Preview>
  <Story name="-10%">
   <Donut progress="-0.1"/>
  </Story>
</Preview>



## Donut with threshold lower than 0%
<Preview>
    <Story name="Threshold under 0%">
       <Donut progress="0.2" warningThreshold="-0.1"/>
     </Story>
</Preview>


## Donut with threshold higher than 100%
<Preview>
    <Story name="Threshold over 100%">
       <Donut progress="1" warningThreshold="1.5"/>
     </Story>
</Preview>           

## Props for Donut

<Props of={Donut} />


