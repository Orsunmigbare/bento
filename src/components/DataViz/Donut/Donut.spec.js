import React from "react"
import renderer from "react-test-renderer"
import { render, screen } from "@testing-library/react"
import { ThemeProvider } from "styled-components"
import baseTheme from "../../../theming/themes/baseTheme"
import "@testing-library/jest-dom"
import Donut from "./Donut"

describe("Donut", () => {
  describe("Snapshots", () => {
    it("No props", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Donut />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("Progress", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Donut progress={0.9} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("Progress&WarningTreshold", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Donut progress={0.7} warningThreshold={0.4} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
  it("Progress over 1", () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={baseTheme}>
          <Donut progress={1.2} />
        </ThemeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("Progress under 1", () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={baseTheme}>
          <Donut progress={-0.9} />
        </ThemeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("invalid progress", () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={baseTheme}>
          <Donut progress={null} />
        </ThemeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("With children", () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={baseTheme}>
          <Donut progress={0.9}>
            <p>whyy?</p>
          </Donut>
        </ThemeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  describe("Rendering", () => {
    it("Should render donut", () => {
      render(
        <ThemeProvider theme={baseTheme}>
          <Donut progress={0.8} />
        </ThemeProvider>
      )
      expect(screen.getByText("80%")).toBeInTheDocument()
    })
  })
})
