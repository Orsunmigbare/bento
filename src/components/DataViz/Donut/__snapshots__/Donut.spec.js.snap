// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Donut Progress over 1 1`] = `
<svg
  className="sc-AxjAm dESvVA"
  height="24"
  viewBox="0 0 24 24"
  width="24"
>
  <title>
    120
    %
  </title>
  <g
    fill="none"
    fillRule="evenodd"
    strokeWidth="6"
  >
    <circle
      className="sc-AxirZ MZXYp"
      cx="12"
      cy="12"
      r="9"
    />
    <circle
      className="sc-AxiKw eqEwrc"
      cx="12"
      cy="12"
      r="9"
      transform="rotate(-90 12 12)"
    />
  </g>
</svg>
`;

exports[`Donut Progress under 1 1`] = `
<svg
  className="sc-AxjAm dESvVA"
  height="24"
  viewBox="0 0 24 24"
  width="24"
>
  <title>
    -90
    %
  </title>
  <g
    fill="none"
    fillRule="evenodd"
    strokeWidth="6"
  >
    <circle
      className="sc-AxirZ MZXYp"
      cx="12"
      cy="12"
      r="9"
    />
    <circle
      className="sc-AxiKw kJmevy"
      cx="12"
      cy="12"
      r="9"
      transform="rotate(-90 12 12)"
    />
  </g>
</svg>
`;

exports[`Donut Snapshots No props 1`] = `
<svg
  className="sc-AxjAm dESvVA"
  height="24"
  viewBox="0 0 24 24"
  width="24"
>
  <title>
    0
    %
  </title>
  <g
    fill="none"
    fillRule="evenodd"
    strokeWidth="6"
  >
    <circle
      className="sc-AxirZ MZXYp"
      cx="12"
      cy="12"
      r="9"
    />
    <circle
      className="sc-AxiKw eRhFxV"
      cx="12"
      cy="12"
      r="9"
      transform="rotate(-90 12 12)"
    />
  </g>
</svg>
`;

exports[`Donut Snapshots Progress 1`] = `
<svg
  className="sc-AxjAm dESvVA"
  height="24"
  viewBox="0 0 24 24"
  width="24"
>
  <title>
    90
    %
  </title>
  <g
    fill="none"
    fillRule="evenodd"
    strokeWidth="6"
  >
    <circle
      className="sc-AxirZ MZXYp"
      cx="12"
      cy="12"
      r="9"
    />
    <circle
      className="sc-AxiKw eZOBHO"
      cx="12"
      cy="12"
      r="9"
      transform="rotate(-90 12 12)"
    />
  </g>
</svg>
`;

exports[`Donut Snapshots Progress&WarningTreshold 1`] = `
<svg
  className="sc-AxjAm dESvVA"
  height="24"
  viewBox="0 0 24 24"
  width="24"
>
  <title>
    70
    %
  </title>
  <g
    fill="none"
    fillRule="evenodd"
    strokeWidth="6"
  >
    <circle
      className="sc-AxirZ MZXYp"
      cx="12"
      cy="12"
      r="9"
    />
    <circle
      className="sc-AxiKw ouiXJ"
      cx="12"
      cy="12"
      r="9"
      transform="rotate(-90 12 12)"
    />
  </g>
</svg>
`;

exports[`Donut With children 1`] = `
<svg
  className="sc-AxjAm dESvVA"
  height="24"
  viewBox="0 0 24 24"
  width="24"
>
  <title>
    90
    %
  </title>
  <g
    fill="none"
    fillRule="evenodd"
    strokeWidth="6"
  >
    <circle
      className="sc-AxirZ MZXYp"
      cx="12"
      cy="12"
      r="9"
    />
    <circle
      className="sc-AxiKw eZOBHO"
      cx="12"
      cy="12"
      r="9"
      transform="rotate(-90 12 12)"
    />
  </g>
</svg>
`;

exports[`Donut invalid progress 1`] = `
<svg
  className="sc-AxjAm dESvVA"
  height="24"
  viewBox="0 0 24 24"
  width="24"
>
  <title>
    0
    %
  </title>
  <g
    fill="none"
    fillRule="evenodd"
    strokeWidth="6"
  >
    <circle
      className="sc-AxirZ MZXYp"
      cx="12"
      cy="12"
      r="9"
    />
    <circle
      className="sc-AxiKw eRhFxV"
      cx="12"
      cy="12"
      r="9"
      transform="rotate(-90 12 12)"
    />
  </g>
</svg>
`;
